topi/python/topi/cuda/conv2d.py

```diff
@@ -27,45 +35,56 @@ def conv2d_cuda(data, kernel, stride, padding, layout='NCHW', out_dtype='float32
     layout : str
         layout of data
 
+    out_dtype: str
+        The output type. This is used for mixed precision.
+
     Returns
     -------
     output : tvm.Tensor
         4-D with shape [batch, out_channel, out_height, out_width]
     """
-    assert isinstance(stride, int) or len(stride) == 2
-    if isinstance(stride, int):
-        stride_h = stride_w = stride
-    else:
-        stride_h, stride_w = stride
-    if isinstance(padding, int):
-        pad_h = pad_w = padding
-    else:
-        pad_h, pad_w = padding
-    # handle dilation
-    dilation_h = dilation_w = 1
-    kernel_tvm = kernel
-    kernel_cudnn = kernel
-    if isinstance(kernel.op, tvm.tensor.ComputeOp) and "dilate" in kernel.op.tag:
-        kernel_before_dilation = kernel.op.input_tensors[0]
-        kernel_cudnn = kernel_before_dilation
-        if layout == 'NCHW':
-            dilation_h = (get_const_int(kernel.shape[2]) + get_const_int(kernel_before_dilation.shape[2]) - 1) \
-                // get_const_int(kernel_before_dilation.shape[2])
-            dilation_w = (get_const_int(kernel.shape[3]) + get_const_int(kernel_before_dilation.shape[3]) - 1) \
-                // get_const_int(kernel_before_dilation.shape[2])
-        elif layout == 'NHWC':
-            dilation_h = (get_const_int(kernel.shape[1]) + get_const_int(kernel_before_dilation.shape[1]) - 1) \
-                // get_const_int(kernel_before_dilation.shape[1])
-            dilation_w = (get_const_int(kernel.shape[2]) + get_const_int(kernel_before_dilation.shape[2]) - 1) \
-                // get_const_int(kernel_before_dilation.shape[2])
     target = tvm.target.current_target()
+
     if "cudnn" in target.libs:
-        assert layout != 'HWCN', "HWCN layout not supported with CUDNN."
-        tensor_format = 0 # CUDNN_TENSOR_NCHW
-        if layout == 'NHWC':
+        if layout == 'NCHW':
+            tensor_format = 0 # CUDNN_TENSOR_NCHW
+            N, _, H, W = get_const_tuple(data.shape)
+        elif layout == 'NHWC':
             tensor_format = 1 # CUDNN_TENSOR_NHWC
+            N, H, W, _ = get_const_tuple(data.shape)
+        else:
+            raise ValueError("Unsupported layout %s in cudnn" % layout)
+        CO, CI, KH, KW = get_const_tuple(kernel.shape)
+
+        # handle dilation
+        stride_h, stride_w = (strides, strides) if isinstance(strides, int) else strides
+        pad_h, pad_w = (padding, padding) if isinstance(padding, int) else padding
+
+        OH = (H + 2 * pad_h - KH) // stride_h + 1
+        OW = (W + 2 * pad_w - KW) // stride_w + 1
+        cfg.add_flop(2 * N * OH * OW * CO * CI * KH * KW)
```

Comment:
this line seems unnecessary, since this is a cudnn call.

Reply:
This line is useful. This makes code consistent so that we can extract FLOP with both cudnn enabled of disabled.